x-pack/plugins/security_solution/server/endpoint/routes/resolver/docs/README.md

```diff
@@ -0,0 +1,214 @@
+3. `A`'s ancestry array is `[3,2]`, query for the lifecycle events for processes with `entity_id` 3 or 2
+4. Check to see if we have retrieved enough ancestors to fulfill the request (we have not, we only received 2 nodes of the 3 that were requested)
+5. We haven't so use the most distant ancestor in our result set (process 2)
+6. Use process 2's ancestry array to query for the next set of results to fulfill the request
+7. Process 2's ancestry array is `[1]` so repeat the process in steps 3-4 and retrieve process with entity_id 1. This fulfills the request so we can return the results for the lifecycle events of A, 3, 2, and 1.
+
+If process 2 had an ancestry array of `[1,0]` we know that we only need 1 more process to fulfill the request so we can truncate the array to `[1]` instead of searching for all the entries in the array.
+
+More generically: In the event where our request stops at the x (non-final) position in an ancestry array, we won't search all items in the array, just those up to the x position. The next-cursor will be set to the last ancestor received since there might be more data.
+
+The `nextAncestor` cursor will be set to `1` in this scenario because we retrieved all 3 ancestors from ES but we don't know if ES has anymore.
+
+## Descendants
+
+We can also leverage the ancestry array to query for the descendants of a process. The basic query for the descendants of a process is: _find all processes where their ancestry array contains a particular entity_id_. The results of this query will be sorted in ascending order by the timestamp field. I will try to outline a couple different scenarios for retrieving descendants using the ancestry array below.
+
+### Start events vs all lifecycle events
+
+There are two parts to querying for descendant process nodes. When a request comes in for 7 process nodes we need to communicate to ES that we want all of the lifecycle nodes for 7 processes. We could use a query that retrieves all lifecycle events (start, end, etc) but the issue with this is that we need to indicate a `size` in our ES query. If we set the `size` to 7, we will only get 7 lifecycle events. These events could be start, end, or already_running events. It doesn't guarantee that we get all of the lifecycle events for 7 process nodes.
+
+Instead we can first query for 7 start events, which guarantees that we will have 7 unique process descendants and then we can gather all those entity_ids and do another query for all the lifecycle events for those 7 processes. The downside here is that you have to do two queries to retrieve all the lifecycle events. Optimizations can be made for the first query for the start events by reducing the `_source` that ES returns to only include the `entity_id` and `ancestry`. This will reduce the amount of data that ES has to send back and speed up the query.
```

Comment:
:grey_question: 

> which guarantees that we will have 7 unique process descendants

Didn't we just learn that this isn't _always_ the case?

Reply:
Good question. So the issue we saw recently was because the endpoint will send a process already_running event (`event.type: ['info']`) when the endpoint finds a process that was spawned before the endpoint started. So if the endpoint is restarted it will send another already_running event.

This is different for `start` and `end` events which I'm referring to here. The endpoint _should_ only send a single `start` and `end` event for a particular entity_id. If a process "restarts" a new unique entity_id will be generated which will not match the previous `start` event.

So when we query for 7 **start** events we will get 7 unique process nodes.

If however the endpoint **did** send a duplicate `start` event (another start event with the same entity_id) it would be a bug. The backend does handle this case though. We use a map to store the results and use the `entity_id` as the key, so if two start events have the same `entity_id` we'll just use the first one and ignore the other.